cloud/shared/bin/lib/log.sh

```diff
@@ -0,0 +1,102 @@
+#! /usr/bin/env bash
+
+#######################################
+# Record a deployment success event in the log.
+# Args:
+#   1: The tag of the deployed server binary.
+#   2: Identifier for the user who initiated the command.
+# Globals read:
+#   LOG_TEMPFILE
+#######################################
+function log::deploy_succeeded() {
+  log::ensure_log_file_fetched
```

Comment:
question (blocking): How do we make sure that the existing log file is up to date? For example, lets say we fetched a log file and while we were appending to it, but before we upload, someone else logged some stuff and uploaded it. and then we'd end up overwriting anything they logged. To minimize that risk, should we make sure to fetch an up-to date log file immediately before any log operation?

Reply:
Valid concern, moreover if two entities are running commands in overlapping timeframes we'll likely have other problems such as inconsistent cloud state.

IMO the canonical solution here is to introduce some manner of expiring mutex lock on the log and therefore the overall cloud state so that if a command is one progress, another can't be run and interfere with the outcome.

There are a few ways we could approach this, but none of them are very simple. I'm inclined to add a section on this to our design doc and revisit it once initial deployments are working.